test/filesystem.cc

```diff
@@ -135,6 +137,70 @@ TEST(Filesystem, extension) {
 TEST(Filesystem, file_size) {
 }
 
+TEST(Filesystem, concurrent_folder_create_delete) {
+
+  const char* create_subdir = "test/a/b/c/d/e/f/g/h/i/j";
+  const char* base_subdir = "test/a";
+
+  auto create_folder = [&]() {
+    bool success = filesystem::create_directories(create_subdir);
+    EXPECT_TRUE(success);
+    EXPECT_TRUE(filesystem::exists(create_subdir));
+  };
+
+  auto remove_folder = [&]() {
+    std::uintmax_t delete_count = filesystem::remove_all(base_subdir);
+    // I've found that two+ threads will claim that they've deleted the same
+    // file object. This results in double+ counting and prevents me from
+    // asserting anything related to the delete_count. All we can be sure
+    // is the following:
+    EXPECT_TRUE(!filesystem::exists(base_subdir));
+  };
+
+  size_t num_threads = std::thread::hardware_concurrency();
+
+  const int num_iters = 50;
+  for (int j = 0; j < num_iters; j++) {
+    // concurrent folder creation
+    {
+      std::vector<std::unique_ptr<std::thread>> threads(num_threads);
+      for (size_t i = 0; i < num_threads; i++)
+        threads[i].reset(new std::thread(create_folder));
+      for (size_t i = 0; i < num_threads; i++)
+        threads[i]->join();
+      EXPECT_TRUE(filesystem::exists(create_subdir));
+    }
+
+    // populate each subfolder with num_files_per_folder text files.
+    {
+      const int num_files_per_folder = 10;
+      std::string folder_name = "test";
+      for (char subdir = 'a'; subdir < 'k'; subdir++) {
+        folder_name = folder_name + "/" + subdir;
+        for (size_t i = 0; i < num_files_per_folder; i++) {
+          std::string filename = folder_name + "/" + std::to_string(i) + ".txt";
+          FILE* f = fopen(filename.c_str(), "w");
```

Comment:
Nit: we should probably use `std::ofstream` here, which is consistent with using the C++ standard library for I/O throughout the project.

Reply:
Probably so.  Y'all are gonna get a kick out of my old school programming.  :)